subsys/bluetooth/host/keys.c

```diff
@@ -335,17 +362,33 @@ static int keys_set(const char *name, size_t len_rd, settings_read_cb read_cb,
 		BT_ERR("Failed to allocate keys for %s", bt_addr_le_str(&addr));
 		return -ENOMEM;
 	}
+	if (len != BT_KEYS_STORAGE_LEN) {
+#if defined(CONFIG_BT_KEYS_OVERWRITE_OLDEST)
+		if (len == (BT_KEYS_STORAGE_LEN_COMPAT)) {
+			BT_WARN("Compatibility loading with no aging counter");
+			memcpy(keys->storage_start, val, len);
+			keys->aging_counter = ++aging_counter_val;
+			last_keys_updated = keys;
+		} else {
+#endif /* CONFIG_BT_KEYS_OVERWRITE_OLDEST */
+			BT_ERR("Invalid key length %zu != %zu", len,
+			       BT_KEYS_STORAGE_LEN);
+			bt_keys_clear(keys);
 
-	if (len_rd != BT_KEYS_STORAGE_LEN) {
-		BT_ERR("Invalid key length %zu != %zu", len,
-		       BT_KEYS_STORAGE_LEN);
-		bt_keys_clear(keys);
-		return -EINVAL;
+			return -EINVAL;
+#if defined(CONFIG_BT_KEYS_OVERWRITE_OLDEST)
+		}
+#endif /* CONFIG_BT_KEYS_OVERWRITE_OLDEST */
```

Comment:
This kind of build-time conditional braces make things quite hard to follow (sorry that I missed it in my earlier review). Since you're returning from the function in the `else` branch, what if you invert the original condition to `if (len != BT_KEYS_STORAGE_LEN_COMPAT)` and return from that branch you don't need an `else` branch at all, and then you have the entire if-branch behind an #ifdef?

Reply:
I still need the return if #ifdef condition is not satisfied. Found another solution.